### DISCUSSION TITLE
[@types/node] v18.14.2: Error instance for `util.promisify(chpro.execFile)`

### DISCUSSION THREAD

**Discussion by thernstig** [DefinitelyTyped/DefinitelyTyped#65754] [Issues with a @types package]
```js
import chpro from 'node:child_process';
const execFile = util.promisify(chpro.execFile);

try {
  await execFile(
    'doesnotexist',
  );
} catch (error) {
  // 'error' is of type 'unknown'.ts(18046)
  console.log(error.stderr);
}
```

How do I properly do a error instanceof X here to so that error is not of type unknown? And so that it contains the stderr?

The relevant file for the type is from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/child_process.d.ts but I cannot deduct from that what to do.

It might be that this typing is missing, but it looks like it is missing support.

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
In `try/catch` we always have `any` or [`unknown` (from TypeScript 4.0)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#unknown-on-catch-clause-bindings). It is a TypeScript limitation, not related to these typings.

You can use some type guards or code like this:
```ts
} catch (error) {
  if (error instanceof Error) {
    console.log((error as ExecFileException).stderr);
  }
}
```

**Reply by thernstig** [reply]
@Semigradsky I do not think this is answered. It was not about `any` or `unknown` in this instance. The problem is that `stderr` is not defined. Look here:

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/30827238/842cfece-81c9-4b64-895b-db40b8d581d3)

As you can see in the linked https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/child_process.d.ts it is not defined anywhere.

**Reply by Semigradsky** [reply]
@thernstig Ok, I don't understant your question =) Could you please provide link to Node.js doc where described than `execFile` error has `stderr` property?

**Reply by thernstig** [reply]
Of course @Semigradsky 😄 

https://nodejs.org/api/child_process.html#child_processexecfilefile-args-options-callback

> If this method is invoked as its [util.promisify()](https://nodejs.org/api/util.html#utilpromisifyoriginal)ed version, it returns a Promise for an Object with stdout and stderr properties. The returned ChildProcess instance is attached to the Promise as a child property. In case of an error (including any error resulting in an exit code other than 0), **a rejected promise is returned, with the same error object given in the callback, but with two additional properties stdout and stderr.**

**Reply by Semigradsky** [reply]
@thernstig here `stdout` and `stderr` are properties of result, not error.
```ts
const { stdout, stderr } = await execFile(
    'doesnotexist',
);
```

**Reply by thernstig** [reply]
@Semigradsky what you wrote is correct, but you missed the second sentence which I highlighted.

 > In case of an error (including any error resulting in an exit code other than 0), a rejected promise is returned, with the same error object given in the callback, but with two additional properties stdout and stderr.

I know for a fact that the error contains a `stderr` property since we are using it to check the stderr in case of a thrown error. And that is also what the description says, that it the error should contain a `stderr` property.

**Reply by SimonSchick** [reply]
Unfortunately in JS you will encounter many base types that have arbitrarily attached properties to them, this appears to be one of those  cases, multiple factors are problematic here:

- TS does currently not allow modeling checked exceptions
- The base error type(?) is having additional properties attached to it

The best you can do is create a function that will check wether an error matches the type:

```ts
interface MyErrorType extends Error {
  stderr: ReadStream;
  stdout: ReadStream;
}

function isMyErrorType(err: Error): err is MyErrorType {
  return 'stderr' in err && 'stdout' in err;
}

try {
 ...
} catch(err) { // unknown
  if (isMyErrorType(err)) {
    // logic
    return;
  }
  throw err
```

**Reply by SimonSchick** [reply]
Explicit example:

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/3259089/1850cab3-4cb8-4af9-8b2d-d340687cc19a)

`err` is of the base type `Error` with bolted on properties, this is not modelable in TS.

**Reply by Semigradsky** [reply]
> you missed the second sentence which I highlighted

Oh, sorry, you are right

**Reply by Semigradsky** [reply]
@pizzacat83 Maybe you can help here?
I see that you [was changing `ExecFileException` lately](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65426). And you have more chance to do needed changes properly.

**Reply by peterblazejewicz** [reply]
the definition ships:
```ts

type ExecFileException =
    & Omit<ExecException, 'code'>
    & Omit<NodeJS.ErrnoException, 'code'>
    & { code?: string | number | undefined | null };
```
and as those types extends Error, which does not have `stderr`, you won't get it atm. If it's legit prop, should be added and PR is welcome.

**Reply by thernstig** [reply]
@peterblazejewicz should this be created/converted into a proper issue then? Who does this? :)

**Reply by peterblazejewicz** [reply]
DT is a fully community driven effort, anyone can do this, thanks!

**Reply by thernstig** [reply]
@peterblazejewicz that I do understand. What I meant is that this is the "Discussions" and not under "Issues" in this repo. I cannot create an actual Issue in this repo it seems.

**Reply by thernstig** [reply]
@Semigradsky @peterblazejewicz who can create an issue out of this?

**Reply by thernstig** [reply]
I can see this turned into an issue here, thanks: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/66093
